### ISSUE TITLE
Mail isn`t send to smart host for only one sender of configured domain

### ISSUE THREAD

**Issue by Vopigro** [MailCleaner/MailCleaner#514]
We configured a smarthost for one of our customer domains. With the except of one particular sender, all the other senders are sending correctly to the configured smarthost. 
The remaining one always sends directly to the server resolved by MX Lookup of recipient.
Exim Version is "Exim version 4.94.2 #2 built 06-May-2021 22:33:17"

**Comment by Vopigro** [comment]
Thank you for your response. 
I testet with Version 4.96 but the error still exists. The Router used for mails referring to stage1 logs is "dnslookup" which for my understanding should only be used if no prior router matches. Neither /etc/exim/mc_binary/mc_router or full_whitelist are in use.

**Comment by JohnMertz** [comment]
You are correct. This means that the smarthost router parameters must not be satisfied:

```
send_to_smart_host:
  driver = manualroute
  transport = remote_smtp
  domains = !+relay_to_domains
  route_data = ${lookup{${domain:$h_from:}}wildlsearch{VARDIR/spool/tmp/mailcleaner/domains_smarthost.list}}
```
Specifically:
- the 'domains' parameter is true so long as the recipient isn't in the 'relay_to_domains' list. ie. If the recipient is a domain which is also hosted by MailCleaner, the smarthost will not be used, it will instead fall through to the next router. This exception is necessary because two local domains both with a smarthost configured could result in a loop.
- the 'route_data' will look up the domain name from the 'From' header and find a hostname (and port) to use within '/var/mailcleaner/spool/tmp/mailcleaner/domains_smarthost.list'. If a match is not found, it will also fall through to the next router.

Neither of these should apply to a specific sender address. However, it is possible that you see an example where that specific sender is trying to deliver to another local address. Can you confirm that the problem exists for external addresses? Say, by testing with a Gmail recipient? Likewise, if this is what is happening, you should be able to replicate it for any sender by sending to another local recipient.

Another possibility for the latter is that this sender is spoofing a different 'From' header domain. If this is the cause, it may be better for us to update that router to check the domain in the envelope sender address, but I would need to deeply investigate and test that possibility to see why we determined that the From header was more appropriate in the first place.

**Comment by Vopigro** [comment]
We have a bunch of mails for external recipients. If I look at the logs for past week we have about 300 outgoing messages from this address to external recipients and only one mail of this sender is handled by the send_to_smart_host router.
I asked the customer to send a mail to my M365 address and this mail as well went through the dnslookup router. I checked the string expansion with 

`/opt/exim4/bin/exim -be '${lookup{${domain:<inserted original header value of From: field here>}}wildlsearch{VARDIR/spool/tmp/mailcleaner/domains_smarthost.list}}'`

and i got the correct host, port back.
Output: 
`hostname_of_smarthost::25 byname randomize`

By design local recipients shouldnd pass our mailcleaner gateway at all and are handled internally. I double checked with the logs and didnt find any mails from this sender to another local address.

**Comment by JohnMertz** [comment]
Glad to hear that you found the issue.

I'm somewhat surprised that this would be a problem though. It seems like a bug in Exim's parsing of the string. Even when decoded, the canonical domain in your example string should still match (assuming that 'senderAddress' is a valid email address with a domain part). A ',' specifically may be an issue because in other headers it delimits an array, but the parser should be smart enough to ignore it in a 'Realname' context.

I can investigate implementing this upstream and perhaps more generally, but I have some concerns about that. Typically the decoded string is what we want to match and there would probably be way more false-negatives due to UTF-8 encoding, etc. We'll need to collect some statistics to see how common either failure case is and ultimately we might need to add logic to catch either `$h_from` or `$rh_from`.

**Comment by Vopigro** [comment]
Yeah I had that concern as well and changed the config to use both implementations of the send_to_smart_host router. 
Once with h_from and if that doesnt match rh_from should also be checked before continuing with the other routers. 

```
send_to_smart_host:
  driver = manualroute
  transport = remote_smtp
  domains = !+relay_to_domains
  route_data = ${lookup{${domain:$h_from:}}wildlsearch{VARDIR/spool/tmp/mailcleaner/domains_smarthost.list}}

send_to_smart_host_new:
  driver = manualroute
  transport = remote_smtp
  domains = !+relay_to_domains
  route_data = ${lookup{${domain:$rh_from:}}wildlsearch{VARDIR/spool/tmp/mailcleaner/domains_smarthost.list}}
```
  
  
With that configuration (currently) all of our testmails pass the correct router.
I'm going to check the log files in the next couple of days to see if everything works fine.
